mudslide/turbomole_model.py

```diff
@@ -0,0 +1,234 @@
+        states: ArrayLike,
+        sub_dir_stem:str = "traj",
+        representation: str = "adiabatic",
+        reference: Any = None,
+        turbomole_modules = {"gs_energy": "ridft", "gs_grads": "rdgrad", "es_grads": "egrad"}
+    ):
+        ElectronicModel_.__init__(self, representation=representation, reference=reference)
+
+        self.turbomole_dir = turbomole_dir
+        self.states = states
+        self.nstates_ = len(self.states)
+        
+        self.sub_dir_stem = sub_dir_stem
+        self.sub_dir_num = 0
+
+        self.turbomole_modules = turbomole_modules
+        self.turbomole_init()
+
+        assert turbomole_is_installed()
+        assert all([ shutil.which(x) is not None for x in self.turbomole_modules.values() ])
+
+    
+    def nstates(self):
+            return self.nstates_
+
+    def ndim(self):
+            return self.ndim_
+
+    def turbomole_init(self):
+        self.coord_path = Path(self.turbomole_dir)/"control"  
+        self.get_coords()
+
+    def get_coords(self):
+        coords = []
+        self.atom_order = []
+        self.mass = []
+
+        coord = subprocess.run(["sdg", "coord"], capture_output= True, text = True).stdout
+    
+        coord_list = coord.rstrip().split("\n")
+
+        for c in coord_list[1:]:
+            c_list = c.split()
+            self.atom_order.append(c_list[3])
+            coords.append([float(val) for val in c_list[:3]])
+            self.mass.append(3 * [masses[c_list[3]]])
+
+        self.ndim_ = 3 * len(coords)
+        self.X = np.array(coords, dtype=np.float64).reshape(self.ndim())
+        self.mass = np.array(self.mass, dtype=np.float64).reshape(self.ndim()) * amu_to_au
+
+
+    def update_coords(self, X):
+        self.coord_path = subprocess.run(["sdg", "-f", "coord"], capture_output=True, text=True).stdout.rstrip()
+        X = X.reshape((self.ndim() // 3, 3))
+        with open(self.coord_path, "r") as f:
+            lines = f.readlines()
+        regex = re.compile(r"\$coord\n|\$coord\s")
+        read_start = np.argwhere(np.array([bool(regex.match(line)) for line in lines]))[0][0]
+
+        current_line = read_start +1
+        for i, coord_list in enumerate(X):
+            lines[current_line] = "    {: .14f}      {: .14f}      {: .14f}      {}\n".format(
+                                  coord_list[0], coord_list[1], coord_list[2], self.atom_order[i]
+                    )
+            current_line += 1
+        
+        with open(self.coord_path, "w") as coord_file:
+            coord_file.write("".join(lines))            
+
+
+    def call_turbomole(self, outname="turbo.out"):
+        # Open file/run Turbomole
+        with open(outname, "w") as f:
+            for turbomole_module in self.turbomole_modules.values():
+                tur_output = subprocess.run(turbomole_module, stdout=f) 
+
+        # Parse results with Turboparse
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
```

Comment:
this isn't doing anything... Are you expecting it to?

Reply:
Hmm I think this part of the code was repeated - deleted!